### PR TITLE
Dedup allowances in crypto approve allowance transaction handler

### LINKED ISSUE #3662: Importer failed to persist nft allowance
### Description

Stable testnet importer threw exception when ingesting nft allowances:

```
Apr 27 13:23:05 mirrornode-stable-testnet-primary-parser-1 java[11739]: 2022-04-27T07:23:05.490-0600 ERROR scheduling-5 c.h.m.i.p.r.RecordFileParser Error parsing file 2022-04-27T13_06_53.142509000Z.rcd after 10.16 ms com.hedera.mirror.importer.exception.ParserException: Error copying 5 items to table nft_allowance_temp
        at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:112)
        at com.hedera.mirror.importer.parser.batch.CompositeBatchPersister.persist(CompositeBatchPersister.java:62)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.executeBatches(SqlEntityListener.java:259)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:186)
        at com.hedera.mirror.importer.parser.record.entity.sql.SqlEntityListener.onEnd(SqlEntityListener.java:76)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:135)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.doParse(RecordFileParser.java:48)
        at com.hedera.mirror.importer.parser.AbstractStreamFileParser.parse(AbstractStreamFileParser.java:79)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:111)
        at com.hedera.mirror.importer.parser.record.RecordFileParser.parse(RecordFileParser.java:48)
        at com.hedera.mirror.importer.parser.record.RecordFileParser$$FastClassBySpringCGLIB$$8b9310db.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:783)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753)
        at org.springframework.retry.interceptor.RetryOperationsInterceptor$1.doWithRetry(RetryOperationsInterceptor.java:97)
        at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:329)
        at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:209)
        at org.springframework.retry.interceptor.RetryOperationsInterceptor.invoke(RetryOperationsInterceptor.java:133)
        at org.springframework.retry.annotation.AnnotationAwareRetryOperationsInterceptor.invoke(AnnotationAwareRetryOperationsInterceptor.java:163)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:698)
        at com.hedera.mirror.importer.parser.record.RecordFileParser$$EnhancerBySpringCGLIB$$fd079491.parse(<generated>)
        at com.hedera.mirror.importer.config.MessagingConfiguration.lambda$integrationFlow$1(MessagingConfiguration.java:107)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.springframework.integration.handler.LambdaMessageProcessor.processMessage(LambdaMessageProcessor.java:104)
        at org.springframework.integration.handler.ServiceActivatingHandler.handleRequestMessage(ServiceActivatingHandler.java:105)
        at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:136)
        at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:56)
        at org.springframework.integration.endpoint.PollingConsumer.handleMessage(PollingConsumer.java:158)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.messageReceived(AbstractPollingEndpoint.java:475)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.doPoll(AbstractPollingEndpoint.java:461)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.pollForMessage(AbstractPollingEndpoint.java:413)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$4(AbstractPollingEndpoint.java:348)
        at org.springframework.integration.util.ErrorHandlingTaskExecutor.lambda$execute$0(ErrorHandlingTaskExecutor.java:57)
        at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50)
        at org.springframework.integration.util.ErrorHandlingTaskExecutor.execute(ErrorHandlingTaskExecutor.java:55)
        at org.springframework.integration.endpoint.AbstractPollingEndpoint.lambda$createPoller$5(AbstractPollingEndpoint.java:341)
        at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
        at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:95)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.hedera.mirror.importer.exception.ParserException: Error copying 5 items to table nft_allowance_temp
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:111)
        at com.hedera.mirror.importer.parser.batch.BatchInserter.persist(BatchInserter.java:108)
        ... 54 more
Caused by: org.postgresql.util.PSQLException: ERROR: ON CONFLICT DO UPDATE command cannot affect row a second time
  Hint: Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:134)
        at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertToFinalTable(BatchUpserter.java:117)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.insertItems(BatchUpserter.java:160)
        at com.hedera.mirror.importer.parser.batch.BatchUpserter.persistItems(BatchUpserter.java:103)
        ... 55 more
```

Though v0.55.0 was able to process the record file in the log since it process nft serial allowance differently, the issue still exists that if there are duplicate entries in *Allowance in a crypto approve allowance transaction, the timestamp for the duplicates but the last one will be `empty` and the upsert sql will try to insert all such rows into db, which evetually causes the  "ON CONFLICT DO UPDATE command cannot affect row a second time" error.

### Steps to reproduce

refer to the description

### Additional context

_No response_

### Hedera network

testnet

### Version

v0.55.0

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by xin-hedera**
**Description**:

This PR fixes the bug exposed in stable testnet by dedup allowances in crypto approve allowance transaction handler.

**Related issue(s)**:

Fixes #3662

**Notes for reviewer**:

Services changed to allow duplicate allowances in a crypto approve allowance transaction. The duplicate allowances can cause mirirornode importer to violate postgresql "ON CONFLICT DO UPDATE cannot affect row a second time" constraint.

The duplicate allowances result in rows in the temp table where the first `n - 1` have an empty `timestamp_range` and the last allowance has an open-ended `timestamp_range`, the upsert sql then will try to insert all such rows to the current table (e.g., `nft_allowance`), thus it violates the postgresql constraint.
 
**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListenerCryptoTest.java [hunk 1]
```diff
@@ -1053,6 +1053,10 @@ private List<NftAllowance> customizeNftAllowances(Timestamp consensusTimestamp,
                 .setSpender(spender2)
                 .setTokenId(tokenId)
                 .build());
+
+        // duplicate nft allowance
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListenerCryptoTest.java:1057]: right. trying to improve the test case with minimal changes

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/CryptoApproveAllowanceTransactionHandler.java [hunk 2]
```diff
@@ -84,11 +96,27 @@ public void updateTransaction(Transaction transaction, RecordItem recordItem) {
...
             cryptoAllowance.setPayerAccountId(payerAccountId);
             cryptoAllowance.setSpender(EntityId.of(cryptoApproval.getSpender()).getId());
-            cryptoAllowance.setTimestampLower(consensusTimestamp);
-            entityListener.onCryptoAllowance(cryptoAllowance);
+            cryptoAllowance.setTimestampLower(recordItem.getConsensusTimestamp());
+
+            if (!cryptoAllowanceState.containsKey(cryptoAllowance.getId())) {
+                entityListener.onCryptoAllowance(cryptoAllowance);
+                cryptoAllowanceState.put(cryptoAllowance.getId(), cryptoAllowance);
+            }
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/CryptoApproveAllowanceTransactionHandler.java:104]: Could optimize to only search the map once.

```suggestion
            if (cryptoAllowanceState.putIfAbsent(cryptoAllowance.getId(), cryptoAllowance) == null) {
                entityListener.onCryptoAllowance(cryptoAllowance);
            }
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/CryptoApproveAllowanceTransactionHandler.java [hunk 3]
```diff
@@ -84,11 +96,27 @@ public void updateTransaction(Transaction transaction, RecordItem recordItem) {
             cryptoAllowance.setOwner(ownerAccountId.getId());
             cryptoAllowance.setPayerAccountId(payerAccountId);
             cryptoAllowance.setSpender(EntityId.of(cryptoApproval.getSpender()).getId());
-            cryptoAllowance.setTimestampLower(consensusTimestamp);
-            entityListener.onCryptoAllowance(cryptoAllowance);
+            cryptoAllowance.setTimestampLower(recordItem.getConsensusTimestamp());
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/CryptoApproveAllowanceTransactionHandler.java:99]: Could pull up to a variable outside the loop like payerAccountId

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListenerCryptoTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListenerCryptoTest.java:1057]: I think this simple check is fine here. I think you should add more exhaustive tests at the `CryptoApproveAllowanceTransactionHandlerTest` level. We should be wary about adding exhaustive tests at the EntityRecordItemListenerTests as they are more complicated and slower to run. Prefer unit tests over integration tests.

[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/entity/EntityRecordItemListenerCryptoTest.java:1057]: updated `CryptoApproveAllowanceTransactionHandlerTest` to verify the last such allowance is honored

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by Nana-EC — APPROVED** [review]
LGTM

**Pull request merged**